Fix package/dependency creation bugs

AyanSinhaMahapatra opened the pull request:
<!-- Delete Template sections if unneccesary -->
<!-- Add issue number here (We encourage you to create the Issue First) -->
<!-- You can also link the issue in Commit Messages -->

<!--
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!
* [x] - Checked Box
* [ ] - Unchecked Box
-->

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

<!--
We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged**
if your tests fail at first!
If tests do fail, click on the red `X` to learn why by reading the logs.
Thanks!
-->

<!-- Don't forget to Signoff -->

JonoYang committed:
Assign npm Resources to Package #2929

Signed-off-by: Jono Yang <jyang@nexb.com>

AyanSinhaMahapatra committed:
Fix infinite recursion bug for pypi and others

Fixes #2928 in pypi and others which were being caused by infinite recursion between the
functions `assign_package_to_parent_tree` and `assign_package_to_resources` because of
improper overriding.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix package creation crash for purl less deps

Fixes #2911

Do not create top level dependencies in case of dependencies
without purls, these are encountered in pypi requirements
with local paths.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge remote-tracking branch 'origin/2929-npm-for-packages' into fix-2911-purl-less-deps

pombredanne commented on src/packagedcode/pypi.py:
```diff
@@ -71,6 +71,14 @@ class BasePypiHandler(models.DatafileHandler):
     def compute_normalized_license(cls, package):
         return compute_normalized_license(package.declared_license)
 
+    @classmethod
```
What do you gain from overriding these two methods with the exact same code that's in the super class?

AyanSinhaMahapatra replied:
If in pypi the `PythonEditableInstallationPkgInfoFile.assign_package_to_resources` is overridden to call `cls.assign_package_to_parent_tree` then, the `assign_package_to_parent_tree` again calls the `assign_package_to_resources` function which has now been overridden to call the same. This causes the recursion.

pombredanne reviewed (approved):
LGTM!
I just have a small comment wrt. https://github.com/nexB/scancode-toolkit/pull/2932/files#r861907921 where subclassing is not needed especially since you call super() which will travel up the class hierarchy and MRO alright

AyanSinhaMahapatra committed:
Call functions from models.DatafileHandler

Call resource updating functions directly from models.DatafileHandler
instead of calling super().

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@pombredanne updated that.

AyanSinhaMahapatra committed:
Do not fail in package/deps creation

Try and catch failures in top level packages and
dependencies creation, and add them to resource level
`scan_errors`, instead of crashing.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix test failures

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne committed:
Merge remote-tracking branch 'upstream/develop' into fix-2911-purl-less-deps

pombredanne committed:
Add missing .gitignored files

with git add --force

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Also un-ignore pip.egg-info test directory

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Adapt test

There is still a pending bug.
See: https://github.com/nexB/scancode-toolkit/issues/2934

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne commented:
LGTM. Thanks! merging!
